Order placement verification

I'm trying to place a buy and sell order for WETH/BTC. I see at least two issues here:
```
2019/05/01 05:51:27 [DEBUG] Placing order: {0.0305 350000000 WETH/BTC  BUY} 
2019/05/01 05:51:27 [ERROR] Could not place orders: {0.0305 350000000 0 WETH/BTC} - rpc error: code = Unknown desc = WETH outbound balance of 0 is not sufficient for order amount of 35
00000000000000000
```
- Since this is a BUY order, I would like to get WETH and pay with BTC. For that I don't need to have WETH outbound balance.
- This order should be prevented since I don't have any Raiden channel. 

```
2019/05/01 05:51:27 [DEBUG] Placing order: {0.0315 350000000 WETH/BTC  SELL} 
2019/05/01 05:51:27 [DEBUG] Order added: price:0.0315 quantity:350000000 pair_id:"WETH/BTC" id:"29384370-6bd5-11e9-b365-0db8c15a4f61" local_id:"29384370-6bd5-11e9-b365-0db8c15a4f61" cr
eated_at:1556689887527 side:SELL is_own_order:true 
```
- for this order I do need WETH balance (which I don't have).
- Should not be able to open this order as I don't have raiden channel.

Was done on xud1

--- a/lib/orderbook/OrderBook.ts
+++ b/lib/orderbook/OrderBook.ts
@@ -332,16 +332,15 @@ class OrderBook extends EventEmitter {
       };
     }
 
-    if (!this.nosanitychecks) {
+    if (!this.nosanitychecks && this.swaps) {
       // check if sufficient outbound channel capacity exists
-      const { makerAmount } = Swaps.calculateSwapAmounts(order.quantity, order.price, order.isBuy, order.pairId);
-      const { makerCurrency } = Swaps.deriveCurrencies(order.pairId, order.isBuy);
-      const swapClient = this.swaps && this.swaps.swapClients.get(makerCurrency);
+      const { outboundCurrency, outboundAmount } = Swaps.calculateInboundOutboundAmounts(order.quantity, order.price, order.isBuy, order.pairId);
+      const swapClient = this.swaps.swapClients.get(outboundCurrency);
       if (!swapClient) {
-        throw errors.SWAP_CLIENT_NOT_FOUND(makerCurrency);
+        throw errors.SWAP_CLIENT_NOT_FOUND(outboundCurrency);
       }
-      if (makerAmount > swapClient.maximumOutboundCapacity) {
-        throw errors.INSUFFICIENT_OUTBOUND_BALANCE(makerCurrency, makerAmount, swapClient.maximumOutboundCapacity);
+      if (outboundAmount > swapClient.maximumOutboundCapacity) {
+        throw errors.INSUFFICIENT_OUTBOUND_BALANCE(outboundCurrency, outboundAmount, swapClient.maximumOutboundCapacity);
       }
     }
 

--- a/lib/swaps/Swaps.ts
+++ b/lib/swaps/Swaps.ts
@@ -59,20 +59,6 @@ class Swaps extends EventEmitter {
   }
 
   /**
-   * Derives the maker and taker currency for a swap.
-   * @param pairId The trading pair id for the swap
-   * @param isBuy Whether the maker order in the swap is a buy
-   * @returns An object with the derived `makerCurrency` and `takerCurrency` values
-   */
-  public static deriveCurrencies = (pairId: string, isBuy: boolean) => {
-    const [baseCurrency, quoteCurrency] = pairId.split('/');
-
-    const makerCurrency = isBuy ? baseCurrency : quoteCurrency;
-    const takerCurrency = isBuy ? quoteCurrency : baseCurrency;
-    return { makerCurrency, takerCurrency };
-  }
-
-  /**
    * Checks if a swap request is valid. This is a shallow check that only detects critical
    * inconsistencies and verifies only whether the request can possibly lead to a successful swap.
    * @returns `true` if the request is valid, otherwise `false`
@@ -84,19 +70,40 @@ class Swaps extends EventEmitter {
   }
 
   /**
-   * Calculates the amount of subunits/satoshis each side of a swap should receive.
+   * Calculates the currencies and amounts of subunits/satoshis each side of a swap should receive.
    * @param quantity The quantity being swapped
    * @param price The price for the swap
    * @param isBuy Whether the maker order in the swap is a buy
-   * @returns An object with the calculated `makerAmount` and `takerAmount` values
+   * @returns An object with the calculated maker and taker values.
    */
-  public static calculateSwapAmounts = (quantity: number, price: number, isBuy: boolean, pairId: string) => {
+  private static calculateMakerTakerAmounts = (quantity: number, price: number, isBuy: boolean, pairId: string) => {
+    const { inboundCurrency, inboundAmount, outboundCurrency, outboundAmount } =
+      Swaps.calculateInboundOutboundAmounts(quantity, price, isBuy, pairId);
+    return {
+      makerCurrency: inboundCurrency,
+      makerAmount: inboundAmount,
+      takerCurrency: outboundCurrency,
+      takerAmount: outboundAmount,
+    };
+  }
+
+  /**
+   * Calculates the incoming and outgoing currencies and amounts of subunits/satoshis for an order if it is swapped.
+   * @param quantity The quantity of the order
+   * @param price The price of the order
+   * @param isBuy Whether the order is a buy
+   * @returns An object with the calculated incoming and outgoing values.
+   */
+  public static calculateInboundOutboundAmounts = (quantity: number, price: number, isBuy: boolean, pairId: string) => {
     const [baseCurrency, quoteCurrency] = pairId.split('/');
     const baseCurrencyAmount = Math.round(quantity * Swaps.UNITS_PER_CURRENCY[baseCurrency]);
     const quoteCurrencyAmount = Math.round(quantity * price * Swaps.UNITS_PER_CURRENCY[quoteCurrency]);
-    const makerAmount = isBuy ? baseCurrencyAmount : quoteCurrencyAmount;
-    const takerAmount = isBuy ? quoteCurrencyAmount : baseCurrencyAmount;
-    return { makerAmount, takerAmount };
+
+    const inboundCurrency = isBuy ? baseCurrency : quoteCurrency;
+    const inboundAmount = isBuy ? baseCurrencyAmount : quoteCurrencyAmount;
+    const outboundCurrency = isBuy ? quoteCurrency : baseCurrency;
+    const outboundAmount = isBuy ? quoteCurrencyAmount : baseCurrencyAmount;
+    return { inboundCurrency, inboundAmount, outboundCurrency, outboundAmount };
   }
 
   public init = async () => {
@@ -188,8 +195,7 @@ class Swaps extends EventEmitter {
       throw SwapFailureReason.SwapClientNotSetup;
     }
 
-    const { makerCurrency } = Swaps.deriveCurrencies(maker.pairId, maker.isBuy);
-    const { makerAmount } = Swaps.calculateSwapAmounts(taker.quantity, maker.price, maker.isBuy, maker.pairId);
+    const { makerCurrency, makerAmount } = Swaps.calculateMakerTakerAmounts(taker.quantity, maker.price, maker.isBuy, maker.pairId);
 
     const swapClient = this.swapClients.get(makerCurrency)!;
 
@@ -292,10 +298,9 @@ class Swaps extends EventEmitter {
   private beginSwap = async (maker: PeerOrder, taker: OwnOrder): Promise<string> => {
     const peer = this.pool.getPeer(maker.peerPubKey);
 
-    const { makerCurrency, takerCurrency } = Swaps.deriveCurrencies(maker.pairId, maker.isBuy);
-
     const quantity = Math.min(maker.quantity, taker.quantity);
-    const { makerAmount, takerAmount } = Swaps.calculateSwapAmounts(quantity, maker.price, maker.isBuy, maker.pairId);
+    const { makerCurrency, makerAmount, takerCurrency, takerAmount } =
+      Swaps.calculateMakerTakerAmounts(quantity, maker.price, maker.isBuy, maker.pairId);
     const clientType = this.swapClients.get(makerCurrency)!.type;
     const destination = peer.getIdentifier(clientType, makerCurrency)!;
 
@@ -362,8 +367,7 @@ class Swaps extends EventEmitter {
 
     const { quantity, price, isBuy } = orderToAccept;
 
-    const { makerAmount, takerAmount } = Swaps.calculateSwapAmounts(quantity, price, isBuy, requestBody.pairId);
-    const { makerCurrency, takerCurrency } = Swaps.deriveCurrencies(requestBody.pairId, isBuy);
+    const { makerCurrency, makerAmount, takerCurrency, takerAmount } = Swaps.calculateMakerTakerAmounts(quantity, price, isBuy, requestBody.pairId);
 
     const swapClient = this.swapClients.get(takerCurrency);
     if (!swapClient) {
@@ -493,7 +497,7 @@ class Swaps extends EventEmitter {
         // TODO: penalize peer
         return;
       } else if (quantity < deal.proposedQuantity) {
-        const { makerAmount, takerAmount } = Swaps.calculateSwapAmounts(quantity, deal.price, deal.isBuy, deal.pairId);
+        const { makerAmount, takerAmount } = Swaps.calculateMakerTakerAmounts(quantity, deal.price, deal.isBuy, deal.pairId);
         deal.takerAmount = takerAmount;
         deal.makerAmount = makerAmount;
       }

--- a/test/simulation/tests.go
+++ b/test/simulation/tests.go
@@ -158,8 +158,8 @@ func testOrderBroadcastAndInvalidation(net *xudtest.NetworkHarness, ht *harnessT
 	ht.act.verifyConnectivity(net.Alice, net.Bob)
 
 	req := &xudrpc.PlaceOrderRequest{
-		Price:    10,
-		Quantity: 100000000,
+		Price:    0.02,
+		Quantity: 10000000,
 		PairId:   "LTC/BTC",
 		OrderId:  "random_order_id",
 		Side:     xudrpc.OrderSide_BUY,
@@ -180,7 +180,7 @@ func testOrderMatchingAndSwap(net *xudtest.NetworkHarness, ht *harnessTest) {
 	// Place an order on Alice.
 	req := &xudrpc.PlaceOrderRequest{
 		OrderId:  "maker_order_id",
-		Price:    10,
+		Price:    0.02,
 		Quantity: 10000000,
 		PairId:   "LTC/BTC",
 		Side:     xudrpc.OrderSide_BUY,

--- a/test/unit/Swaps.spec.ts
+++ b/test/unit/Swaps.spec.ts
@@ -54,6 +54,14 @@ describe('Swaps', () => {
     makerAmount: buyDeal.takerAmount,
   };
 
+  const buyOrder = {
+    quantity,
+    price,
+    pairId,
+    isBuy: true,
+  };
+  const sellOrder = { ...buyOrder, isBuy: false };
+
   const swapRequest: SwapRequestPacketBody = {
     takerCltvDelta,
     orderId,
@@ -62,40 +70,52 @@ describe('Swaps', () => {
     pairId: 'LTC/BTC',
   };
 
-  it('should derive currencies for a buy order', () => {
-    const { makerCurrency, takerCurrency } = Swaps['deriveCurrencies'](buyDeal.pairId, buyDeal.isBuy);
-    expect(makerCurrency).to.equal(buyDeal.makerCurrency);
-    expect(takerCurrency).to.equal(buyDeal.takerCurrency);
-  });
-
-  it('should derive currencies for a sell order', () => {
-    const { makerCurrency, takerCurrency } = Swaps['deriveCurrencies'](sellDeal.pairId, sellDeal.isBuy);
-    expect(makerCurrency).to.equal(sellDeal.makerCurrency);
-    expect(takerCurrency).to.equal(sellDeal.takerCurrency);
-  });
-
-  it('should calculate swap amounts for a buy order', () => {
-    const { makerAmount, takerAmount } = Swaps['calculateSwapAmounts'](
+  it('should calculate swap amounts and currencies for a buy order', () => {
+    const { makerCurrency, makerAmount, takerCurrency, takerAmount } = Swaps['calculateMakerTakerAmounts'](
       buyDeal.quantity!, buyDeal.price, buyDeal.isBuy, buyDeal.pairId,
     );
     expect(makerAmount).to.equal(buyDeal.makerAmount);
     expect(takerAmount).to.equal(buyDeal.takerAmount);
+    expect(makerCurrency).to.equal(buyDeal.makerCurrency);
+    expect(takerCurrency).to.equal(buyDeal.takerCurrency);
   });
 
-  it('should calculate swap amounts for a sell order', () => {
-    const { makerAmount, takerAmount } = Swaps['calculateSwapAmounts'](
+  it('should calculate swap amounts and currencies for a sell order', () => {
+    const { makerCurrency, makerAmount, takerCurrency, takerAmount } = Swaps['calculateMakerTakerAmounts'](
       sellDeal.quantity!, sellDeal.price, sellDeal.isBuy, sellDeal.pairId,
     );
     expect(makerAmount).to.equal(sellDeal.makerAmount);
     expect(takerAmount).to.equal(sellDeal.takerAmount);
+    expect(makerCurrency).to.equal(sellDeal.makerCurrency);
+    expect(takerCurrency).to.equal(sellDeal.takerCurrency);
   });
 
-  it('should calculate swap amounts for a WETH buy order', () => {
-    const { makerAmount, takerAmount } = Swaps['calculateSwapAmounts'](
+  it('should calculate swap amounts and currencies for a WETH buy order', () => {
+    const { makerCurrency, makerAmount, takerCurrency, takerAmount } = Swaps['calculateMakerTakerAmounts'](
       buyDealEth.quantity!, buyDealEth.price, buyDealEth.isBuy, buyDealEth.pairId,
     );
     expect(makerAmount).to.equal(buyDealEth.makerAmount);
     expect(takerAmount).to.equal(buyDealEth.takerAmount);
+    expect(makerCurrency).to.equal(buyDealEth.makerCurrency);
+    expect(takerCurrency).to.equal(buyDealEth.takerCurrency);
+  });
+
+  it('should calculate inbound and outbound amounts and currencies for a buy order', () => {
+    const { inboundCurrency, inboundAmount, outboundCurrency, outboundAmount } =
+      Swaps.calculateInboundOutboundAmounts(quantity, price, true, pairId);
+    expect(inboundCurrency).to.equal('LTC');
+    expect(inboundAmount).to.equal(Swaps['UNITS_PER_CURRENCY']['LTC'] * quantity);
+    expect(outboundCurrency).to.equal('BTC');
+    expect(outboundAmount).to.equal(Swaps['UNITS_PER_CURRENCY']['LTC'] * quantity * price);
+  });
+
+  it('should calculate inbound and outbound amounts and currencies for a sell order', () => {
+    const { inboundCurrency, inboundAmount, outboundCurrency, outboundAmount } =
+      Swaps.calculateInboundOutboundAmounts(quantity, price, false, pairId);
+    expect(inboundCurrency).to.equal('BTC');
+    expect(inboundAmount).to.equal(Swaps['UNITS_PER_CURRENCY']['LTC'] * quantity * price);
+    expect(outboundCurrency).to.equal('LTC');
+    expect(outboundAmount).to.equal(Swaps['UNITS_PER_CURRENCY']['LTC'] * quantity);
   });
 
   it(`should validate a good swap request`, () => {

